### PR TITLE
Add integration tests for publishing to CloudWatch logs

### PR DESCRIPTION
**Pull request by SaxyPandaBear**
# Description of the issue
Currently no integration tests in the GitHub repo that validate that the agent publishes logs to CloudWatch Logs

# Description of changes
1. integration test that publishes all logs from the file to CWL
2. integration feature test for agent-side log filtering (#327)
3. emits `AWS_REGION` to linux terraform so that the runner can use the AWS SDK 
4. Update README for setting up integration tests

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
ran on my personal fork.
```
aws_instance.integration-test (remote-exec): === RUN   TestWriteLogsToCloudWatch
aws_instance.integration-test (remote-exec): 2022/04/08 15:42:45 Found instance id i-02a8dcd645c29596c
aws_instance.integration-test (remote-exec): === RUN   TestWriteLogsToCloudWatch/Happy_path
aws_instance.integration-test (remote-exec): 2022/04/08 15:42:45 Copy File resources/config_log.json to /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/08 15:42:45 File resources/config_log.json abs path /home/ec2-user/amazon-cloudwatch-agent/integration/test/cloudwatchlogs/resources/config_log.json
aws_instance.integration-test (remote-exec): 2022/04/08 15:42:45 File : resources/config_log.json copied to : /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test: Still creating... [6m0s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/08 15:42:50 Writing 100 of each log type
aws_instance.integration-test: Still creating... [6m10s elapsed]
aws_instance.integration-test: Still creating... [6m20s elapsed]
aws_instance.integration-test: Still creating... [6m30s elapsed]
aws_instance.integration-test: Still creating... [6m40s elapsed]
aws_instance.integration-test: Still creating... [6m50s elapsed]
aws_instance.integration-test: Still creating... [7m0s elapsed]
aws_instance.integration-test: Still creating... [7m10s elapsed]
aws_instance.integration-test: Still creating... [7m20s elapsed]
aws_instance.integration-test: Still creating... [7m30s elapsed]
aws_instance.integration-test: Still creating... [7m40s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/08 15:44:30 Finished writing logs. Sleeping before starting agent...
aws_instance.integration-test: Still creating... [7m50s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/08 15:44:41 Agent has started
aws_instance.integration-test: Still creating... [8m0s elapsed]
aws_instance.integration-test: Still creating... [8m10s elapsed]
aws_instance.integration-test: Still creating... [8m20s elapsed]
aws_instance.integration-test: Still creating... [8m30s elapsed]
aws_instance.integration-test: Still creating... [8m40s elapsed]
aws_instance.integration-test: Still creating... [8m50s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/08 15:45:41 Agent is stopped
aws_instance.integration-test (remote-exec): 2022/04/08 15:45:41 Checking i-02a8dcd645c29596c/i-02a8dcd645c29596c since 2022-04-08 15:42:45.637383387 +0000 UTC m=+0.006182571 for 200 expected logs
aws_instance.integration-test (remote-exec): === RUN   TestWriteLogsToCloudWatch/Client-side_log_filtering
aws_instance.integration-test (remote-exec): 2022/04/08 15:45:41 Copy File resources/config_log_filter.json to /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/08 15:45:41 File resources/config_log_filter.json abs path /home/ec2-user/amazon-cloudwatch-agent/integration/test/cloudwatchlogs/resources/config_log_filter.json
aws_instance.integration-test (remote-exec): 2022/04/08 15:45:41 File : resources/config_log_filter.json copied to : /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/08 15:45:46 Writing 100 of each log type
aws_instance.integration-test: Still creating... [9m0s elapsed]
aws_instance.integration-test: Still creating... [9m10s elapsed]
aws_instance.integration-test: Still creating... [9m20s elapsed]
aws_instance.integration-test: Still creating... [9m30s elapsed]
aws_instance.integration-test: Still creating... [9m40s elapsed]
aws_instance.integration-test: Still creating... [9m50s elapsed]
aws_instance.integration-test: Still creating... [10m0s elapsed]
aws_instance.integration-test: Still creating... [10m10s elapsed]
aws_instance.integration-test: Still creating... [10m20s elapsed]
aws_instance.integration-test: Still creating... [10m30s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/08 15:47:26 Finished writing logs. Sleeping before starting agent...
aws_instance.integration-test: Still creating... [10m40s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/08 15:47:37 Agent has started
aws_instance.integration-test: Still creating... [10m50s elapsed]
aws_instance.integration-test: Still creating... [11m0s elapsed]
aws_instance.integration-test: Still creating... [11m10s elapsed]
aws_instance.integration-test: Still creating... [11m20s elapsed]
aws_instance.integration-test: Still creating... [11m30s elapsed]
aws_instance.integration-test: Still creating... [11m40s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/08 15:48:37 Agent is stopped
aws_instance.integration-test (remote-exec): 2022/04/08 15:48:37 Checking i-02a8dcd645c29596c/i-02a8dcd645c29596c since 2022-04-08 15:45:41.559547245 +0000 UTC m=+175.928346429 for 100 expected logs
aws_instance.integration-test (remote-exec): --- PASS: TestWriteLogsToCloudWatch (351.77s)
aws_instance.integration-test (remote-exec):     --- PASS: TestWriteLogsToCloudWatch/Happy_path (175.92s)
aws_instance.integration-test (remote-exec):     --- PASS: TestWriteLogsToCloudWatch/Client-side_log_filtering (175.80s)
aws_instance.integration-test (remote-exec): PASS
aws_instance.integration-test (remote-exec): ok  	github.com/aws/amazon-cloudwatch-agent/integration/test/cloudwatchlogs	351.782s
```

### CODE REVIEW

File: integration/test/cloudwatchlogs/publish_logs_test.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	c, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		// fail fast so we don't continue the test
+		t.Fatalf("Error occurred while creating SDK config: %v", err)
+	}
+	client := imds.NewFromConfig(c)
+	metadata, err := client.GetInstanceIdentityDocument(ctx, &imds.GetInstanceIdentityDocumentInput{})
+	if err != nil {
+		t.Fatalf("Error occurred while retrieving EC2 instance ID: %v", err)
+	}
```
Comment (adam-mateen) [inline integration/test/cloudwatchlogs/publish_logs_test.go:62]: Could use `assert`.
```assert.NilError(t, err, "blah %v", err)```

Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/publish_logs_test.go:62]: I agree, normally, but as I've learned in doing this, doing something like `assert.NoError(t, err)` does not fail fast. So after the test gets an error trying to read from IMDS, it will continue and then panic. In the grand scheme of things, it's still a test so it probably doesn't mean much harm either way, but basically what was happening was I was seeing errors about trying to access/reference nil further down in the tests, which is an immediate red herring to the root of the issue. `t.Fatal()` logs the message and then calls `t.FailNow()`, which pulls the plug on the test. 

That being said, if you still feel strongly about using `assert` in this scenario, I can switch it back. I do realize that me writing code like this is contrary to most of my previous review comments related to these types of assertions 😃

File: integration/test/cloudwatchlogs/resources/write_logs.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+if [ -z "$2" ]; then
+  file_path='/tmp/test.log'
+fi
+
+echo "Emitting $n logs to $file_path"
+
+for i in $(seq 1 $n); do
+  echo "$(date -Is) [foo] This is a log line" >>$file_path
+  echo "$(date -Is) [bar] This is a log line" >>$file_path
+  sleep 1 # sleep so the timestamps for the logs are different
```
Comment (ZhenyuTan-amz) [inline integration/test/cloudwatchlogs/resources/write_logs.sh:20]: Just to make sure this is 1ms right? If 1s I feel 100 iteration is a bit overkill.

File: integration/test/cloudwatchlogs/resources/write_logs.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+done
```
Comment (adam-mateen) [inline integration/test/cloudwatchlogs/resources/write_logs.sh:21]: You could delete this shell script and write it in Golang.
In general I would recommend sticking with the primary programming language of the code repo if/when possible.
Then you could use Golang asserts on opening and writing the file.
Another reason would be potentially reusing the same integration test on Windows EC2 instances.

File: integration/test/cloudwatchlogs/resources/write_logs.sh (hunk 2 above)
Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/resources/write_logs.sh:20]: No, `sleep` is 1 second. 🤔  I went back and forth on this. I could include the count in the log line itself, like `This is the $i log line`. The original intention was to make sure that the log lines were a little unique, and I was relying on the timestamp for that. I'm open to ideas.

File: integration/test/cloudwatchlogs/resources/write_logs.sh (hunk 3 above)
Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/resources/write_logs.sh:21]: That's a good point. This along with @ZhenyuTan-amz 's comment, I think I'll rewrite it. Good call out on propping this up for Windows integration tests. I was thinking of duplicating the script but this is better 😛

File: integration/test/cloudwatchlogs/resources/write_logs.sh (hunk 2 above)
Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/resources/write_logs.sh:20]: see https://github.com/aws/amazon-cloudwatch-agent/pull/436#discussion_r846491204 I'm just going to redo this in Golang. I originally had this as a shell script just cause it was quick and easy to test out on a fresh host, but writing the log lines natively in Golang is probably the better approach.

File: integration/test/cloudwatchlogs/publish_logs_test.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+func writeLogsAndRunAgent(iterations int, runtime time.Duration) {
+	log.Printf("Writing %d of each log type\n", iterations)
+	test.RunShellScript(logScriptPath, strconv.Itoa(iterations)) // write logs before starting the agent
+	log.Println("Finished writing logs. Sleeping before starting agent...")
+	time.Sleep(10 * time.Second)
+
+	test.StartAgent(configOutputPath)
```
Comment (adam-mateen) [inline integration/test/cloudwatchlogs/publish_logs_test.go:89]: It is weird to write logs then start the agent.
I think the typical scenario is the Agent is running and logs files get written.
It is definitely good to test this way too though :)

Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/publish_logs_test.go:89]: The reason I did it this way is because it's more controlled, IMHO. I don't want to mess with timing for logs on the host. One thing I could do by rewriting the log writing in Golang is do it concurrently, and then just use a WaitGroup/lock to stop the agent only after the logs are done writing. But the way I set up this test originally with the shell script to emit logs, I didn't want to let fate decide if the test succeeded in time or not (though 1 minute of run time for the agent is probably plenty).

File: integration/test/cloudwatchlogs/publish_logs_test.go (hunk 1 above)
Comment (adam-mateen) [reply integration/test/cloudwatchlogs/publish_logs_test.go:62]: I specifically looked that up and thought it would fail fast?!
https://pkg.go.dev/gotest.tools/assert#hdr-Assert_and_Check
>Assert() and Check() both accept a Comparison, and fail the test when the comparison fails. The one difference is that Assert() will end the test execution immediately (using t.FailNow()) whereas Check() will fail the test (using t.Fail()), return the value of the comparison, then proceed with the rest of the test case. 

I am fine with keeping it as is now that I know why.

File: integration/test/cloudwatchlogs/publish_logs_test.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	"strconv"
+
+	"testing"
+	"time"
+)
+
+const (
+	configOutputPath = "/opt/aws/amazon-cloudwatch-agent/bin/config.json"
+	logScriptPath    = "resources/write_logs.sh"
+	agentRunTime     = 1 * time.Minute
```
[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/publish_logs_test.go:24]: I took this from an existing integration test 😄  will see if I can trim it down, or confirm this is enough

[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/publish_logs_test.go:24]: Default interval for logs is 5 seconds, according to the default force flush interval for the logs section. https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-Configuration-File-Details.html#CloudWatch-Agent-Configuration-File-Logssection

Changing this to sleep for 10 seconds (2x the interval to ensure consistency in publishing) after the logs are done writing, and also added a comment noting the default

**Comment by SaxyPandaBear** [comment]
New test output:
```
aws_instance.integration-test (remote-exec): === RUN   TestWriteLogsToCloudWatch
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:11 Found instance id i-094bb73bcbd55bd25
aws_instance.integration-test (remote-exec): === RUN   TestWriteLogsToCloudWatch/Happy_path
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:11 Copy File resources/config_log.json to /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:11 File resources/config_log.json abs path /home/ec2-user/amazon-cloudwatch-agent/integration/test/cloudwatchlogs/resources/config_log.json
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:11 File : resources/config_log.json copied to : /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:12 Agent has started
aws_instance.integration-test: Still creating... [6m10s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:22 Writing 200 logs to /tmp/test.log
aws_instance.integration-test: Still creating... [6m20s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:32 Agent is stopped
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:32 Checking i-094bb73bcbd55bd25/i-094bb73bcbd55bd25 since 2022-04-12T01:37:11Z for 200 expected logs
aws_instance.integration-test (remote-exec): === RUN   TestWriteLogsToCloudWatch/Client-side_log_filtering
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:32 Copy File resources/config_log_filter.json to /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:32 File resources/config_log_filter.json abs path /home/ec2-user/amazon-cloudwatch-agent/integration/test/cloudwatchlogs/resources/config_log_filter.json
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:32 File : resources/config_log_filter.json copied to : /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:32 Agent has started
aws_instance.integration-test: Still creating... [6m30s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:42 Writing 200 logs to /tmp/test.log
aws_instance.integration-test: Still creating... [6m40s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:53 Agent is stopped
aws_instance.integration-test (remote-exec): 2022/04/12 01:37:53 Checking i-094bb73bcbd55bd25/i-094bb73bcbd55bd25 since 2022-04-12T01:37:32Z for 100 expected logs
aws_instance.integration-test (remote-exec): --- PASS: TestWriteLogsToCloudWatch (41.46s)
aws_instance.integration-test (remote-exec):     --- PASS: TestWriteLogsToCloudWatch/Happy_path (20.79s)
aws_instance.integration-test (remote-exec):     --- PASS: TestWriteLogsToCloudWatch/Client-side_log_filtering (20.64s)
aws_instance.integration-test (remote-exec): PASS
aws_instance.integration-test (remote-exec): ok  	github.com/aws/amazon-cloudwatch-agent/integration/test/cloudwatchlogs	41.472s
```

File: integration/terraform/ec2/linux/main.tf [hunk 6]
```diff
@@ -21,6 +21,7 @@ resource "aws_instance" "integration-test" {
       "sudo cp original.pem /opt/aws/amazon-cloudwatch-agent/original.pem",
       "sudo cp combine.pem /opt/aws/amazon-cloudwatch-agent/combine.pem",
       "export LOCAL_STACK_HOST_NAME=${var.local_stack_host_name}",
+      "export AWS_REGION=${var.region}",
```
Comment (sethAmazon) [inline integration/terraform/ec2/linux/main.tf:24]: Why is this needed. Will it not default to the region the instance is located?

Comment (SaxyPandaBear) [reply integration/terraform/ec2/linux/main.tf:24]: The SDK needs a region configuration: https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specifying-the-aws-region

The AMI cleaner runs on the GitHub runner, which has the AWS_REGION value set from a previous action, so that works. The integration tests are run on the EC2 instance, so this isn't set at all.

File: integration/test/cloudwatchlogs/publish_logs_test.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+			start := time.Now()
+			var wg sync.WaitGroup
+
+			test.CopyFile(param.configPath, configOutputPath)
+
+			test.StartAgent(configOutputPath)
+
+			time.Sleep(agentRuntime) // make sure that there is enough time between the timestamps
+			wg.Add(1)
+			go writeLogs(t, &wg, logFilePath, param.iterations)
```
[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/publish_logs_test.go:86]: Ah poo. I just realized that I didn't start the agent in here before the `Wait()` call but after the goroutine. Yeah I can simplify this. I think I was doing it one way and then just never moved everything around where it needed to be.

File: integration/test/cwl_util.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+	_, err = cwlClient.DeleteLogGroup(*clientContext, &cloudwatchlogs.DeleteLogGroupInput{
+		LogGroupName: aws.String(logGroupName),
+	})
+	if err != nil && !errors.As(err, &rnf) {
+		log.Printf("Error occurred while deleting log group %s: %v", logGroupName, err)
+	}
+}
+
+func getCloudWatchLogsClient() (*cloudwatchlogs.Client, *context.Context, error) {
```
Comment (sethAmazon) [inline integration/test/cwl_util.go:76]: This will probs be useful for metrics also. You can change the file this is now or we can change it when I write the 30 metrics test when other teams deploy.

File: integration/test/cloudwatchlogs/publish_logs_test.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	// TODO: this only works for EC2 based testing
+	client := imds.NewFromConfig(c)
+	metadata, err := client.GetInstanceIdentityDocument(ctx, &imds.GetInstanceIdentityDocumentInput{})
+	if err != nil {
+		t.Fatalf("Error occurred while retrieving EC2 instance ID: %v", err)
+	}
+	log.Printf("Found instance id %s", metadata.InstanceID)
+	instanceId := metadata.InstanceID
```
Comment (aateeqi) [inline integration/test/cloudwatchlogs/publish_logs_test.go:72]: nitpick: this could go above 71 and can be used in the logging

File: integration/test/cwl_util.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+package test
+
+import (
+	"context"
+	"errors"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs/types"
+	"github.com/aws/aws-sdk-go/aws"
```
Comment (aateeqi) [inline integration/test/cwl_util.go:15]: Probably not a big deal but is there a reason why we can't use v2 for this lib as well?

File: integration/test/agent_util.go [hunk 11]
```diff
@@ -68,17 +68,21 @@ func ReadAgentOutput(d time.Duration) string {
...
-	out, err = exec.Command("bash", "-c", "sudo ./"+path).Output()
+	bashArgs := []string{"-c", "sudo ./"+path}
+	bashArgs = append(bashArgs, args...)
+
+	//out, err = exec.Command("bash", "-c", "sudo ./"+path, args).Output()
+	out, err = exec.Command("bash", bashArgs...).Output()
 
 	if err != nil {
-		log.Fatal(fmt.Sprint(err) + string(out))
+		log.Fatalf("Erorr occurred when executing %s: %s | %s", path, err.Error(), string(out))
```
Comment (aateeqi) [inline integration/test/agent_util.go:85]: typo

File: integration/test/cwl_util.go (hunk 8 above)
Comment (SaxyPandaBear) [reply integration/test/cwl_util.go:76]: Aren't metrics under a different service than CloudWatch logs? We'd be importing from `service/cloudwatch` for that. I figured it warrants its own utility class, unless we want a holistic util for all AWS SDK operations, which I think would just bloat it.

File: integration/test/cwl_util.go (hunk 10 above)
Comment (SaxyPandaBear) [reply integration/test/cwl_util.go:15]: So, this is one I went back and forth on. The SDK uses pointers to values, and `aws` provides helper functions for that, i.e.: `aws.String(logGroupName)`. I could just change it to not use the package, by doing `&logGroupName`. I'm indifferent. There is no equivalent I found in the v2 SDK.

File: integration/test/agent_util.go (hunk 11 above)
Comment (SaxyPandaBear) [reply integration/test/agent_util.go:85]: Not sure why my IDE didn't catch this. Fixed

**Comment by SaxyPandaBear** [comment]
```
aws_instance.integration-test (remote-exec): === RUN   TestWriteLogsToCloudWatch
aws_instance.integration-test (remote-exec): 2022/04/13 00:21:43 Found instance id i-0603c6c76aaf6daa7
aws_instance.integration-test (remote-exec): === RUN   TestWriteLogsToCloudWatch/Happy_path
aws_instance.integration-test (remote-exec): 2022/04/13 00:21:43 Copy File resources/config_log.json to /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/13 00:21:43 File resources/config_log.json abs path /home/ec2-user/amazon-cloudwatch-agent/integration/test/cloudwatchlogs/resources/config_log.json
aws_instance.integration-test (remote-exec): 2022/04/13 00:21:43 File : resources/config_log.json copied to : /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/13 00:21:43 Agent has started
aws_instance.integration-test: Still creating... [6m10s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/13 00:21:53 Writing 200 logs to /tmp/test.log
aws_instance.integration-test: Still creating... [6m20s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/13 00:22:04 Agent is stopped
aws_instance.integration-test (remote-exec): 2022/04/13 00:22:04 Checking i-0603c6c76aaf6daa7/i-0603c6c76aaf6daa7 since 2022-04-13T00:21:43Z for 200 expected logs
aws_instance.integration-test (remote-exec): === RUN   TestWriteLogsToCloudWatch/Client-side_log_filtering
aws_instance.integration-test (remote-exec): 2022/04/13 00:22:04 Copy File resources/config_log_filter.json to /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/13 00:22:04 File resources/config_log_filter.json abs path /home/ec2-user/amazon-cloudwatch-agent/integration/test/cloudwatchlogs/resources/config_log_filter.json
aws_instance.integration-test (remote-exec): 2022/04/13 00:22:04 File : resources/config_log_filter.json copied to : /opt/aws/amazon-cloudwatch-agent/bin/config.json
aws_instance.integration-test (remote-exec): 2022/04/13 00:22:04 Agent has started
aws_instance.integration-test: Still creating... [6m30s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/13 00:22:14 Writing 200 logs to /tmp/test.log
aws_instance.integration-test: Still creating... [6m40s elapsed]
aws_instance.integration-test (remote-exec): 2022/04/13 00:22:24 Agent is stopped
aws_instance.integration-test (remote-exec): 2022/04/13 00:22:24 Checking i-0603c6c76aaf6daa7/i-0603c6c76aaf6daa7 since 2022-04-13T00:22:04Z for 100 expected logs
aws_instance.integration-test (remote-exec): --- PASS: TestWriteLogsToCloudWatch (41.53s)
aws_instance.integration-test (remote-exec):     --- PASS: TestWriteLogsToCloudWatch/Happy_path (20.79s)
aws_instance.integration-test (remote-exec):     --- PASS: TestWriteLogsToCloudWatch/Client-side_log_filtering (20.70s)
aws_instance.integration-test (remote-exec): PASS
aws_instance.integration-test (remote-exec): ok  	github.com/aws/amazon-cloudwatch-agent/integration/test/cloudwatchlogs	41.540s
```

File: integration/test/cloudwatchlogs/publish_logs_test.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+func writeLogs(t *testing.T, filePath string, iterations int) {
+	f, err := os.Create(filePath)
+	if err != nil {
+		t.Fatalf("Error occurred creating log file for writing: %v", err)
+	}
+	defer f.Close()
+
+	log.Printf("Writing %d logs to %s", iterations*len(logLineIds), filePath)
```
Comment (adam-mateen) [inline integration/test/cloudwatchlogs/publish_logs_test.go:101]: `Writing %d logs`
Would it be more accurate to say
`Writing %d lines`
?

Comment (adam-mateen) [inline integration/test/cloudwatchlogs/publish_logs_test.go:101]: `Writing %d logs`
Would it be more accurate to say
`Writing %d lines`
?

File: integration/test/cloudwatchlogs/publish_logs_test.go [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+		ts := time.Now()
+		for _, id := range logLineIds {
+			_, err = f.WriteString(fmt.Sprintf("%s - [%s] This is a log line.\n", ts.Format(time.StampMilli), id))
+			if err != nil {
+				// don't need to fatal error here. if a log line doesn't get written, the count
+				// when validating the log stream should be incorrect and fail there.
+				t.Logf("Error occurred writing log line: %v", err)
+			}
+		}
+		time.Sleep(2 * time.Millisecond)
```
Comment (adam-mateen) [inline integration/test/cloudwatchlogs/publish_logs_test.go:113]: Why sleep here?
The test would still work without it, right?

Comment (adam-mateen) [inline integration/test/cloudwatchlogs/publish_logs_test.go:113]: Why sleep here?
The test would still work without it, right?

File: integration/test/cloudwatchlogs/publish_logs_test.go (hunk 12 above)
Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/publish_logs_test.go:101]: I think GitHub did the same thing to you that it did to me and double submitted the same comment. Resolving this one

Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/publish_logs_test.go:101]: 🤔  Yeah it would be more accurate. I can change it

**Review by adam-mateen — DISMISSED** [review]

File: integration/test/cloudwatchlogs/publish_logs_test.go (hunk 13 above)
Comment (SaxyPandaBear) [reply integration/test/cloudwatchlogs/publish_logs_test.go:113]: Yeah it would. I think to avoid confusion with these sleep timers, I'll just use the `i` value to differentiate between the logs.

**Review by adam-mateen — APPROVED** [review]
